core/chains/evm/txmgr/stuck_tx_detector.go

```diff
@@ -0,0 +1,381 @@
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/common/txmgr"
+	"github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/gas"
+)
+
+type stuckTxDetectorGasEstimator interface {
+	GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, opts ...feetypes.Opt) (fee gas.EvmFee, chainSpecificFeeLimit uint64, err error)
+}
+
+type stuckTxDetectorClient interface {
+	BatchCallContext(ctx context.Context, b []rpc.BatchElem) error
+}
+
+type stuckTxDetectorTxStore interface {
+	FindTxsByStateAndFromAddresses(ctx context.Context, addresses []common.Address, state types.TxState, chainID *big.Int) (txs []*Tx, err error)
+}
+
+type stuckTxDetectorConfig interface {
+	AutoPurgeStuckTxs() bool
+	AutoPurgeThreshold() uint32
+	AutoPurgeMinAttempts() uint32
+	AutoPurgeDetectionApiUrl() *url.URL
+}
+
+type stuckTxDetector struct {
+	lggr      logger.Logger
+	chainID   *big.Int
+	chainType config.ChainType
+	cfg       stuckTxDetectorConfig
+
+	gasEstimator stuckTxDetectorGasEstimator
+	txStore      stuckTxDetectorTxStore
+	chainClient  stuckTxDetectorClient
+	httpClient   *http.Client
+
+	purgeBlockNumLock sync.RWMutex
+	purgeBlockNumMap  map[common.Address]int64 // Tracks the last block num a tx was purged for each from address if the PurgeOverflowTxs feature is enabled
+}
+
+func NewStuckTxDetector(lggr logger.Logger, chainID *big.Int, chainType config.ChainType, cfg stuckTxDetectorConfig, gasEstimator stuckTxDetectorGasEstimator, txStore stuckTxDetectorTxStore, chainClient stuckTxDetectorClient) *stuckTxDetector {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.DisableCompression = true
+	httpClient := &http.Client{Transport: t}
+	return &stuckTxDetector{
+		lggr:             lggr,
+		chainID:          chainID,
+		chainType:        chainType,
+		cfg:              cfg,
+		gasEstimator:     gasEstimator,
+		txStore:          txStore,
+		chainClient:      chainClient,
+		httpClient:       httpClient,
+		purgeBlockNumMap: make(map[common.Address]int64),
+	}
+}
+
+func (d *stuckTxDetector) LoadPurgeBlockNumMap(ctx context.Context, addresses []common.Address) error {
+	if !d.cfg.AutoPurgeStuckTxs() {
+		return nil
+	}
+	d.purgeBlockNumLock.Lock()
+	defer d.purgeBlockNumLock.Unlock()
+	// Ok to reset the map here since this method could be reloaded with a new list of from addresses
+	d.purgeBlockNumMap = make(map[common.Address]int64)
+	for _, address := range addresses {
+		d.purgeBlockNumMap[address] = 0
+	}
+
+	// Find all fatal error transactions to see if any were from previous purges to properly set the map
```

Comment:
I don't quite understand why you need to do all this.
At startup time, just let the map be empty.
If map value for a certain address is empty, means you can purge txs on this address immediately.
When you purge a Tx, then you set the map value.

Reply:
I originally had it that way but was worried that we'd lose expected rate limiting on restart. If we restart right after a tx got purged, we'd immediately purge the next if one was stuck behind the original. If tx get cleared or reaped, there isn't much more we can do but thought I'd at least give best effort.